src/icon.c

```diff
@@ -0,0 +1,183 @@
+        }
+
+        /* absolute path? */
+        if (iconname[0] == '/' || iconname[0] == '~') {
+                pixbuf = get_pixbuf_from_file(iconname);
+        }
+
+        /* search in icon_path */
+        if (!pixbuf) {
+                char *start = settings.icon_path,
+                     *end, *current_folder, *maybe_icon_path;
+                do {
+                        end = strchr(start, ':');
+                        if (!end) end = strchr(settings.icon_path, '\0'); /* end = end of string */
+
+                        current_folder = g_strndup(start, end - start);
+
+                        for (const char **suf = suffixes; *suf; suf++) {
+                                maybe_icon_path = g_strconcat(current_folder, "/", iconname, *suf, NULL);
+                                pixbuf = get_pixbuf_from_file(maybe_icon_path);
+                                g_free(maybe_icon_path);
+
+                                if (pixbuf)
+                                        break;
+                        }
+
+                        g_free(current_folder);
+                        if (pixbuf)
+                                break;
+
+                        start = end + 1;
+                } while (*(end) != '\0');
+        }
+        if (!pixbuf)
+                LOG_W("No icon found for: '%s'", iconname);
+
+        g_free(uri_path);
+        return pixbuf;
+}
+
+GdkPixbuf *get_pixbuf_from_raw_image(const RawImage *raw_image)
+{
+        GdkPixbuf *pixbuf = NULL;
+
+        pixbuf = gdk_pixbuf_new_from_data(raw_image->data,
+                                          GDK_COLORSPACE_RGB,
+                                          raw_image->has_alpha,
+                                          raw_image->bits_per_sample,
+                                          raw_image->width,
+                                          raw_image->height,
+                                          raw_image->rowstride,
+                                          NULL,
+                                          NULL);
+
+        return pixbuf;
+}
+
+cairo_surface_t *icon_get_for_notification(const notification *n)
+{
+        GdkPixbuf *pixbuf;
+
+        if (n->raw_icon)
+                pixbuf = get_pixbuf_from_raw_image(n->raw_icon);
+        else if (n->icon)
+                pixbuf = get_pixbuf_from_icon(n->icon);
+        else
+                return NULL;
+
+        if (!pixbuf)
+                return NULL;
+
+        int w = gdk_pixbuf_get_width(pixbuf);
+        int h = gdk_pixbuf_get_height(pixbuf);
+        int larger = w > h ? w : h;
+        if (settings.max_icon_size && larger > settings.max_icon_size) {
+                GdkPixbuf *scaled;
+                if (w >= h) {
+                        scaled = gdk_pixbuf_scale_simple(pixbuf,
+                                        settings.max_icon_size,
+                                        (int) ((double) settings.max_icon_size / w * h),
```

Comment:
We're parsing an integer to a `double`, to divide it by an `int` and to multiply it via an `int`.

Wouldn't it be more sophisticated, to **multiply** the `int` first and then divide it afterwards. doesn't need any casts and solely consists of integer arithmetic.

-> same for else condition below.